rfd/0202-db-multi-session-mfa.md

```diff
@@ -0,0 +1,247 @@
+
+The multi-session MFA extends [RFD 155 Scoped Webauthn
+Credentials](https://github.com/gravitational/teleport/blob/master/rfd/0155-scoped-webauthn-credentials.md)
+with a new scope for executing database sessions:
+```diff
+// webauthn.proto
+enum ChallengeScope {
+...
+    // Used for 'tsh db exec' and allows reuse. 
+    SCOPE_DATABASE_MULTI_SESSION = 8;
+}
+```
+
+Similar to `SCOPE_ADMIN_ACTION`, the new scope `SCOPE_DATABASE_MULTI_SESSION`
+will allow reuse of the MFA session data until it expires (5 minutes for
+WebAuthn).
+
+The MFA response will be checked upon auth call of `GenerateUserCerts` where
+user requests a TLS user cert with database route. New logic is added to
+`GenerateUserCerts` where the new scope with reuse is allowed only if the role
+set matching the requested database has `roleset.option.require_session_mfa`
+option set to `multi-session`.
+
+The new scope cannot be used for `GenerateUserCerts` for non-database targets.
+And if the MFA response is validated with existing non-reusable `SCOPE_SESSION`,
+the action should always be allowed.
+
+Here is a quick matrix:
+
+| `require_session_mfa` mode | MFA response scope             | Requested Target | Access |
+|----------------------------|--------------------------------|------------------|--------|
+| `multi-session`            | `SCOPE_SESSION`                | Database         | Allow  |
+| `multi-session`            | `SCOPE_DATABASE_MULTI_SESSION` | Database         | Allow  |
+| `multi-session`            | `SCOPE_DATABASE_MULTI_SESSION` | Non-Database     | Denied |
+| `session`                  | `SCOPE_SESSION`                | Database         | Allow  |
+| `session`                  | `SCOPE_DATABASE_MULTI_SESSION` | Database         | Denied |
+| `session`                  | `SCOPE_DATABASE_MULTI_SESSION` | Non-Database     | Denied |
+
+MFA requirement check is also updated to indicate whether the client can get
+away with reusing the MFA response.
+```diff
+// IsMFARequiredResponse is a response for MFA requirement check.
+message IsMFARequiredResponse {
+// Required is a simplified view over [MFARequired].
+bool Required = 1;
+// MFARequired informs whether MFA is required to access the corresponding
+// resource.
+MFARequired MFARequired = 2;
++// AllowReuse indicates whether an MFA challenge response can be used
++// to authenticate the user more than once until the challenge expires.
++bool allow_reuse = 3;
+}
+```
+
+### The `tsh db exec` command
+
+General flow of the command:
+- Fetch databases (either specified directly or through search).
+- For each database:
+  - Prompt MFA:
+    - If `per-session` MFA is required.
+    - Or, if first time requesting MFA for `multi-session` or shared `multi-session`
+      response is expired.
+    - No MFA if not required.
+  - Starts a local proxy in tunnel mode for this database (regardless of cluster
+    proxy listener mode).
+  - Craft a command for `os.exec`. Outputs are printed to `stdout` unless
```

Comment:
Would stdout output be prefixes with the name of the database that produced the entry?

Reply:
i initially did that in my POC but i find it difficult to copy the output so I dropped the prefix. In cases queries are executed in parallel, we could require the output dir. WDYT?